### PR TITLE
Add codeql action

### PR DESCRIPTION
**Pull request by drmcghee**
**What this PR does / why we need it**:
Scans the code (and tests) in the eraser project for vulnerabilities and errors. 
This PR adds the pull request git hub action for CodeQL within the test.yaml "unit tests"

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes # 176 

**Special notes for your reviewer**:
Did some basic testing, can continue to use local runner for git hub action if a specific scenario
Needs permissions for git hub action to work

### CODE REVIEW

**Comment by sozercan** [comment]
@drmcghee looks like it has some errors running it

**Comment by sozercan** [comment]
@drmcghee are you interested in continuing to work on this PR?

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -57,6 +58,12 @@ jobs:
         uses: actions/checkout@v3
       - name: Unit test
         run: make test
+      - name: Initialize CodeQL
+        uses: github/codeql-action/init@v2
+        with:
+          languages: go
+      - name: Perform CodeQL Analysis
```
Comment (sozercan) [inline .github/workflows/test.yaml:81]: we need an autobuild step here. example: https://github.com/kubernetes/release/blob/master/.github/workflows/codeql-analysis.yml#L65

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -57,6 +58,12 @@ jobs:
         uses: actions/checkout@v3
       - name: Unit test
         run: make test
+      - name: Initialize CodeQL
```
Comment (sozercan) [inline .github/workflows/test.yaml:75]: can we add a parallel job for codeql? seems to take a long time

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -57,6 +58,12 @@ jobs:
...
+        uses: github/codeql-action/init@v2
+        with:
+          languages: go
+      - name: Perform CodeQL Analysis
+        uses: github/codeql-action/analyze@v2
```
Comment (sozercan) [inline .github/workflows/test.yaml]: nit

```suggestion
        uses: github/codeql-action/analyze@v2
```

File: .github/workflows/test.yaml (hunk 1 above)
Comment (drmcghee) [reply .github/workflows/test.yaml:81]: Sorry Sertac, was travelling.. Was travelling. Will review all below - thank you

File: .github/workflows/test.yaml (hunk 2 above)
Comment (drmcghee) [reply .github/workflows/test.yaml:75]: I can try with a parallel job on my machine. 

The samples point to init, make, then analyse.
If a do it in parallel then then perhaps  I can skip the unit tests (make tests)

File: .github/workflows/test.yaml (hunk 1 above)
Comment (drmcghee) [reply .github/workflows/test.yaml:81]: Seems to work on my machine (12 mins for complete run - one vulnerability warning

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -99,47 +127,10 @@ jobs:
...
-      - uses: crazy-max/ghaction-github-runtime@v2
+      - uses: crazy-max/ghaction-github-runtime@v1
       - name: Build eraser-manager
         run: make docker-build-manager CACHE_FROM=type=gha,scope=eraser-manager CACHE_TO=type=gha,scope=eraser-manager,mode=max
       - name: Build eraser
         run: make docker-build-eraser CACHE_FROM=type=gha,scope=eraser-node CACHE_TO=type=gha,scope=eraser-node,mode=max
       - name: Run e2e test
         run: make e2e-test
-
-  scan_vulnerabilities:
```
Comment (sozercan) [inline .github/workflows/test.yaml]: i don't think we want to remove this

File: .github/workflows/test.yaml [hunk 5]
```diff
@@ -99,47 +127,10 @@ jobs:
           path: |
             ~/go/pkg/mod
             ~/.cache/go-build
-      - uses: crazy-max/ghaction-github-runtime@v2
```
Comment (sozercan) [inline .github/workflows/test.yaml]: let's revert these changes

File: .github/workflows/test.yaml [hunk 6]
```diff
@@ -58,17 +57,46 @@ jobs:
...
+  codeql:
+    name: "CodeQL"
+    runs-on: ubuntu-latest
+    permissions: write-all
+    timeout-minutes: 15      
+    steps:
+      - name: Set up Go 1.18
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.18
```
Comment (sozercan) [inline .github/workflows/test.yaml:72]: let's align on 1.17 for now

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -58,17 +57,46 @@ jobs:
...
+      - uses: actions/cache@v3.0.2
```
Comment (sozercan) [inline .github/workflows/test.yaml]: does cache here help with build?

**Review by sozercan — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yaml (hunk 4 above)
Comment (drmcghee) [reply .github/workflows/test.yaml]: I didn't honest governor !.. will add back in

File: .github/workflows/test.yaml (hunk 6 above)
Comment (drmcghee) [reply .github/workflows/test.yaml:72]: ok will revert. I was following a change previous owner applied.

File: .github/workflows/test.yaml (hunk 5 above)
Comment (drmcghee) [reply .github/workflows/test.yaml]: looks like was changed by someone a few days ago

File: .github/workflows/test.yaml (hunk 7 above)
Comment (drmcghee) [reply .github/workflows/test.yaml]: unclear, only if components are being reused. I suspect its more useful for unit tests so I will remove from CodeQL and test

File: .github/workflows/test.yaml (hunk 6 above)
Comment (drmcghee) [reply .github/workflows/test.yaml:72]: current version is 1.18: https://github.com/Azure/eraser/blob/main/.github/workflows/test.yaml

**Review by sozercan — APPROVED** [review]
